runtime/compare.c

```diff
@@ -194,14 +228,30 @@ static intnat do_compare_val(struct compare_stack* stk,
         continue;
     }
     case String_tag: {
-      mlsize_t len1, len2;
+      mlsize_t len1, len2, len_remaining;
       int res;
+      const char *str1, *str2;
+      bool last_block;
       if (v1 == v2) break;
       len1 = caml_string_length(v1);
       len2 = caml_string_length(v2);
-      res = memcmp(String_val(v1), String_val(v2), len1 <= len2 ? len1 : len2);
-      if (res < 0) return LESS;
-      if (res > 0) return GREATER;
+      len_remaining = len1 <= len2 ? len1 : len2;
+      str1 = String_val(v1);
+      str2 = String_val(v2);
+      while (1) {
+        last_block = len_remaining <= COMPARE_SIGNAL_POLL_PERIOD;
+        res = memcmp(str1, str2,
+                     last_block
+                     ? len_remaining
+                     : COMPARE_SIGNAL_POLL_PERIOD);
+        if (res < 0) return LESS;
+        if (res > 0) return GREATER;
+        if (last_block) break;
+        COMPARE_DO_POLL();
```

Comment:
I would expect `COMPARE_INC_POLL_TIMER` here, not `COMPARE_DO_POLL`.

Reply:
It was going to be a `COMPARE_INC_POLL_TIMER(COMPARE_SIGNAL_POLL_PERIOD)` since that's how many bytes were checked, but of course then the comparison would always hold. I've changed it to that, since it makes intention clearer, and the compiler should be able to optimise it away.